parla/task_runtime.py

```diff
@@ -1167,55 +1157,80 @@ def fill_curr_mapped_task_queue(self):
             if len(new_tasks) > 0:
                 self._mapped_task_queue.extendleft(new_tasks)
 
+    def _construct_datamove_task(self, target_data, compute_task):
+        """
+          This function constructs data movement task for target data.
+          This function consists of two steps.
+          First, it iterates all operand data of the dependency tasks
+          of the computation task (original task).
+          If any of the dependency tasks' data is overlapped with the
+          target data, then add the dependency task to the new data
+          movement task's dependency list.
+          Second, construct a data movement task.
+        """
+        dep_for_datamovement_task = []
+        is_overlapped = False
+        for dep in compute_task.dependencies:
+            if not isinstance(dep, Task):
+                continue
+            # Iterate data movement tasks of the compute_task.
+            for dep_data in dep.dataflow:
+                if id(target_data) == id(dep_data):
+                    # If any data of the dependent tasks is
+                    # overlapped with the target data,
+                    # add the corresponding computation task
+                    # to the data movement task's dependency list.
```

Comment:
I think it is necessary. Yes, the data movement tasks should be done before the computation tasks start and after the dependencies start. But the point that I want to make here is that the data movements do not need to wait for all the original tasks dependencies. Each of data movement task has individual dependency. For example, in the above case, C-c does not need to wait for B. C-b needs to wait for both A and B. If C-c just moves data after B is completed, then it would be slower if C-c array is much bigger than C-b.

Reply:
Is current data movement tasks run sync or asynchronously or doesn't matter? Since the current PArray move data without using Stream, cupy.copy() will sync. Does this matter or not?

Reply:
Among tasks, they are asynchronous. My first hand thought is that it should not be matter since parallel data movement task moves different data block. Also, since each datamovement task already runs on its own stream, there is no need to use separate stream on PArray. One concern is not your code, but CuPy copy. As I reported before, and left on your PR, it could use its own stream, which is buggy for our Parla. But we'll see whether it works or not.